pandas/core/internals/managers.py

```diff
@@ -2143,13 +2144,17 @@ def _grouping_func(tup: tuple[int, ArrayLike]) -> tuple[int, bool, DtypeObj]:
     return sep, isinstance(dtype, np.dtype), dtype
 
 
-def _form_blocks(arrays: list[ArrayLike], consolidate: bool) -> list[Block]:
+def _form_blocks(arrays: list[ArrayLike], consolidate: bool, refs: list) -> list[Block]:
     tuples = list(enumerate(arrays))
 
     if not consolidate:
-        nbs = _tuples_to_blocks_no_consolidate(tuples)
+        nbs = _tuples_to_blocks_no_consolidate(tuples, refs)
         return nbs
 
+    # when consolidating, we can ignore refs (either stacking always copies,
```

Comment:
Can we optimise this in a follow up? If we have only one block for a dtype we could avoid copying?

Reply:
> If we have only one block for a dtype we could avoid copying?

Yeah, that has been the long standing behaviour to _always_ copy when consolidate=True, even if not necessary. But indeed when CoW is enabled, we could optimize this.

Reply:
Ok, but lets do in a follow up, not here obviously